azurerm/resource_arm_container_group.go

```diff
@@ -0,0 +1,317 @@
+)
+
+func resourceArmContainerGroup() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmContainerGroupCreate,
+		Read:   resourceArmContainerGroupRead,
+		Update: resourceArmContainerGroupCreate,
+		Delete: resourceArmContainerGroupDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"location": locationSchema(),
+
+			"resource_group_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"ip_address_type": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+				ValidateFunc: validation.StringInSlice([]string{
+					"Public",
+				}, true),
+				Default: "Public",
+			},
+
+			"os_type": {
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+				ValidateFunc: validation.StringInSlice([]string{
+					"windows",
+					"linux",
+				}, true),
+			},
+
+			"tags": tagsSchema(),
+
+			"ip_address": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"container": {
+				Type:     schema.TypeList,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"image": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"cpu": {
+							Type:     schema.TypeFloat,
+							Required: true,
+						},
+
+						"memory": {
+							Type:     schema.TypeFloat,
+							Required: true,
+						},
+
+						"port": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							ValidateFunc: validation.IntBetween(1, 65535),
+						},
+
+						"protocol": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								"tcp",
+								"udp",
+							}, true),
+						},
+
+						"env_var": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"name": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+
+									"value": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceArmContainerGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient)
+	containerGroupsClient := client.containerGroupsClient
+
+	// container group properties
+	resGroup := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+	location := d.Get("location").(string)
+	OSType := d.Get("os_type").(string)
+	IPAddressType := d.Get("ip_address_type").(string)
+	tags := d.Get("tags").(map[string]interface{})
+
+	containersConfig := d.Get("container").([]interface{})
```

Comment:
can we pull all of this out into a separate function `expandContainerGroupContainers` which returns the deserialized object?

Reply:
Done.